MB-2321 Allow revoking individual user sessions

monfresh opened the pull request:
**Description**:
In order to obtain an ATO (Authority to Operate) for MilMove, we need
to provide a way to revoke individual user sessions. Currently,
session management is provided by JWTs per ADR 15, but JWTs aren't
designed to be revoked on an individual basis. Instead, we need to
store session data on the server.

In this PR, we have chosen Redis because it automatically deletes
expired sessions. With Postgres, we would need to run a routine
periodically to clean up stale sessions.

After researching various session management solutions, I chose
`scs` because it was the easiest to integrate, and it supports various
stores out of the box. It is the second most popular repo after
`gorilla/sessions`. I didn't pick `gorilla/sessions` because it suffers
from memory leak issues, and uses its own `context` instead of the
`request.Context()` provided by Golang. The maintainers are aware of
the issues and have opened a GitHub issue to propose improvements for
v2. However, it doesn't look like any progress has been made over the
past 2 years, while `scs` has implemented most of those improvements.

Note that this PR does not provide the ability to limit concurrent
sessions or revoke individual sessions via the admin interface. That
work is being done in a separate branch.

**Setup**:
`docker pull redis`

**Reviewer Notes**:
Things to test:

**milmovelocal auth**
1. Go to milmovelocal:3000
- [ ] Verify that a session cookie named "mil_session_token" is present
(Developer Tools -> Application tab -> Cookies (under Storage in the
left sidebar))
- [ ] Verify that the value in the `Expires/Max-Age` column is `Session`
- [ ] Verify that the HttpOnly column is checked
- [ ] Verify that the Path is `/`
- [ ] Verify that `SameSite` is `Lax`
2. In your Terminal, run `redis-dev`, then type `KEYS *`
- [ ] Verify there is an entry labeled `scs:session:token`, where
`token` is the `Value` of the `mil_session_token` cookie
3. Sign in
- [ ] Verify that after successful sign in, the `Value` of the
`mil_session_token` cookie changes
4. Run `KEYS *` again in the Redis console
- [ ] Verify that the previous entry is gone and that a new one
corresponding to the new session cookie is present
5. Sign out
- [ ] Verify that the previous entry in Redis is gone and that a new one
corresponding to the new session cookie is present
- [ ] Verify that the session cookie changed in the browser
6. In `serve.go`, on line 504, change the IdleTimeout from 15 minutes
to 1 minute
7. Sign in, then wait a little over a minute
8. Try to make a new request without refreshing the browser, for
example, filling out the moves form and clicking the Next button
- [ ] Verify that you are not able to make a request and that you see
an Unauthorized Error. Ideally, the user would be redirected to the
sign in page. I'm working on implementing that.

**devlocal auth**
- [ ] Verify you can sign in and out via devlocal auth flow:
http://milmovelocal:3000/devlocal-auth/login
- [ ] Verify you can create a New milmove User
- [ ] Verify you can create a New dps User

**Role based auth**
1. In your `.envrc.local`, add
`export FEATURE_FLAG_ROLE_BASED_AUTH=true`
2. Stop the server, run `direnv allow`
3. run `echo $FEATURE_FLAG_ROLE_BASED_AUTH` to make sure it's `true`
4. run `make server_run`
5. Go to milmovelocal:3000 and make sure you can sign in and out

**Logging into multiple apps at the same time**
1. Log in to http://milmovelocal:3000/
2. In a separate tab, log in to http://officelocal:3000
3. In a separate tab, log in to http://adminlocal:3000
4. Refresh each tab. Verify that you are still signed into each app.

**References**:
[gorilla sessions issues](https://github.com/gorilla/sessions/issues/105)
[scs repo](https://github.com/alexedwards/scs)

chrisgilmerproj commented:
This looks good! I still need to run this but I appreciate the way you laid out the code. The one thing I'm not seeing is a way to revoke a session without logging into Redis directly. Can you help me understand how I would revoke a session for a user without the redis CLI?

monfresh commented on cmd/milmove/serve.go:
```diff
@@ -477,10 +482,45 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+	pool := &redis.Pool{
+		MaxIdle: 10,
+		Dial: func() (redis.Conn, error) {
+			return redis.Dial("tcp", "localhost:6379")
```
Where should the Redis host be stored? In an ENV var?

chrisgilmerproj replied:
Yeah, environment variables. Make one in `pkg/cli/redis.go` and include the redis hostname and redis port seprately.

monfresh commented on cmd/milmove/serve.go:
```diff
@@ -477,10 +482,45 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+	// This preserves the behavior we had when we managed our own session
+	// cookies, where we extended the session expiry by 15 minutes on every
+	// request while the session was still valid.
+	sessionManager.IdleTimeout = 15 * time.Minute
```
Should this be in an ENV var, so we can change it in prod if necessary without requiring a code change and deploy?

chrisgilmerproj replied:
Yeah, that would be wonderful! Maybe an env var in `pkg/cli/session.go`

monfresh commented on pkg/handlers/internalapi/service_members.go:
```diff
@@ -148,10 +150,12 @@ func (h CreateServiceMemberHandler) Handle(params servicememberop.CreateServiceM
 		session.LastName = *(newServiceMember.LastName)
 	}
+	// Update session cookie here instead of in responders?
+	// h.sessionManager.Put(ctx, "session", session)
```
To preserve the previous behavior, I am updating the cookie in the responder, but I'm wondering if we can do it here instead and not have to use the responder. What is the purpose of the responder?

chrisgilmerproj replied:
I have no idea what the responder is for. Sorry I can't be more help.

monfresh commented:
> This looks good! I still need to run this but I appreciate the way you laid out the code. The one thing I'm not seeing is a way to revoke a session without logging into Redis directly. Can you help me understand how I would revoke a session for a user without the redis CLI?

Currently, you can only do it via the Redis CLI. What do you have in mind for something outside the CLI? Adding something to the GUI admin website?

chrisgilmerproj commented:
> > The one thing I'm not seeing is a way to revoke a session without logging into Redis directly. Can you help me understand how I would revoke a session for a user without the redis CLI?
> 
> Currently, you can only do it via the Redis CLI. What do you have in mind for something outside the CLI? Adding something to the GUI admin website?

Exactly this. In the administrative app (admin.move.mil) we ought to have a way to revoke a session. There's no way for anyone to use the redis CLI in our infrastructure, which is the same for Postgres. So the only way to revoke a session would need to be via a UI in the app. And ideally it would be something like a list of active users where we could click a button next to their name that says "revoke session". It doesn't have to be in this PR but the security finding won't be closed until we have the capability to revoke the session.

monfresh commented:
Gotcha. OK, I'll open a new JIRA ticket for that and work on it in a separate PR.

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -477,10 +495,48 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+			return redis.DialURL(fullRedisURL)
+		},
+	}
+	sessionManager.Store = redisstore.New(pool)
```
In `pkg/cli/dbconn.go` is where we initialize the postgres DB (not a great location to be sure). But moving the creation of the redis pool out of `serve.go` is a good idea, which means we can create interfaces and tests outside of this file.

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -477,10 +495,48 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+	redisURL := v.GetString(cli.RedisURLFlag)
+	redisPort := v.GetString(cli.RedisPortFlag)
+	redisDB := v.GetString(cli.RedisDBFlag)
+	fullRedisURL := fmt.Sprintf("%s%s%s%s%s", redisURL, ":", redisPort, "/", redisDB)
```
I suggest doing a `log.Info` here like we do with the postgres connection, which makes it a lot easier to debug bad URIs in production.

chrisgilmerproj commented on pkg/auth/authentication/auth_test.go:
```diff
@@ -396,77 +428,19 @@ func (suite *AuthSuite) TestAuthorizeDeactivateOfficeUser() {
 	suite.Equal(http.StatusForbidden, rr.Code, "authorizer did not recognize deactivated office user")
 }
 
-func (suite *AuthSuite) TestRedirectLoginGovErrorMsg() {
```
What happened to this test?

monfresh replied:
oops. Adding it back now. Good catch!

chrisgilmerproj commented on pkg/auth/authentication/authghc.go:
```diff
@@ -137,7 +137,6 @@ func (uua UnknownUserAuthorizer) AuthorizeUnknownUser(openIDUser goth.User, sess
 			return err
 		}
 	}
-	uua.logger.Info("logged in", zap.Any("session", session))
```
We might have a reason for logging when folks log into the system. We should ensure there isn't a STIG reason before we remove this.

monfresh replied:
I moved this out of here and into `auth.go` under `AuthorizeUnknownUserNew` on line 534 to be consistent with the other logging of the "logged in" event.

chrisgilmerproj replied:
aha! Thanks.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,40 @@
+
+	if r.MatchString(RedisURLFlag) == false {
+		return errors.Errorf("%s must follow the scheme 'redis://username:password@hostname'", RedisURLFlag)
+	}
```
We should also check the port is correct. Look at `pkg/cli/ports.go` and copy that out.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,40 @@
+// InitRedisFlags initializes RedisFlags command line flags
+func InitRedisFlags(flag *pflag.FlagSet) {
+	flag.String(RedisURLFlag, "redis://localhost", "Redis URL with username, password, and hostname")
+	flag.String(RedisPortFlag, "6379", "Redis port")
```
This should be an integer.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,40 @@
+func InitRedisFlags(flag *pflag.FlagSet) {
+	flag.String(RedisURLFlag, "redis://localhost", "Redis URL with username, password, and hostname")
+	flag.String(RedisPortFlag, "6379", "Redis port")
+	flag.String(RedisDBFlag, "0", "Redis database")
```
What is this flag for? Maybe add more context about the values. And if its a boolean then let's use `flag.Bool()`.

monfresh replied:
This is to specify the DB to use within the same Redis instance, but I'm not sure if that's how we'll be using Redis in prod. The idea is that if we need to store several different things in Redis, we want to keep the sessions data separate from other data. Would we be spinning up separate instances or using the same Redis instance?

chrisgilmerproj replied:
At the moment I expect that we'll be using one redis instance and one redis DB for everything. I still think its worth having this so we can configure it in the future if we need to. I guess the word `database` is just over-used and it would be nice to have more clarity.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,40 @@
+
+// InitRedisFlags initializes RedisFlags command line flags
+func InitRedisFlags(flag *pflag.FlagSet) {
+	flag.String(RedisURLFlag, "redis://localhost", "Redis URL with username, password, and hostname")
```
I'd really like to use `REDIS_USER`, `REDIS_PASSWORD`, `REDIS_HOST` like we do for the postgres DB. And then construct the URI from that.

chrisgilmerproj commented on pkg/cli/session.go:
```diff
@@ -0,0 +1,31 @@
+
+// CheckSession validates session command line flags
+func CheckSession(v *viper.Viper) error {
+	return nil
```
Does the duration always have to be a positive duration? If so we should check for that. And is there an upper limit on duration? Maybe we want to error if someone tries to set the session duration to more than an hour? I would definitely put some guard rails around the different duration here.

monfresh replied:
IdleTimeout doesn't have to be positive. If it's zero or a negative number, it will mean there is no inactivity timeout. What I think I will do here is check if the environment is production, and if so, enforce a range for the idle timeout, say between 15 and 60 minutes. For all other envs I don't think we need to have guard rails. Thoughts?

chrisgilmerproj replied:
I suggest we always have guard rails so that folks know how they work both in and out of AWS. This allows us to test our assumptions in development before we ever get to production.

monfresh replied:
I agree in general. In this case, we want different settings for development versus production. For example, in development, we want people to be able to disable idle timeout completely, which was a feature we used to have before via `NO_SESSION_TIMEOUT`. This means allowing people to set `SESSION_IDLE_TIMEOUT_IN_MINUTES` to 0 or a negative number. In production, however, we want to enforce an inactivity timeout, most likely not less than 15 minutes, and not more than 1 hour.

Similarly, for the session lifetime (absolute expiry that gets set at sign in and doesn't change), it must be greater than 0 or else you won't be able to sign in, so in development, it requires it to be at least 1, but in production, we probably want at least a few hours. If we have a way to measure average session lengths, we can use that to inform the minimum we set.

I verified the production logic in development by changing `EnvironmentProd` to `EnvironmentDevelopment` in this commit: https://github.com/transcom/mymove/pull/3926/commits/e91f62b7e42a15cd345e70ed69ac11aaefd3626d

chrisgilmerproj replied:
There are examples of enforcing rules differently based on environment. Devlocal-auth is one of those things that checks that the devlocal-auth system is only deployed in Experimental. As much as possible codify these rules and make them explicit. It makes it easier to test our assumptions later and it also is inherently documentation for us to read later.

monfresh replied:
I'm not sure if you wrote this comment before seeing this commit, which I think codifies the rules: https://github.com/transcom/mymove/commit/e91f62b7e42a15cd345e70ed69ac11aaefd3626d

monfresh commented:
Another thing I noticed is that the `scs` package uses this function to generate the token:
```golang
func generateToken() (string, error) {
    b := make([]byte, 32)
    _, err := rand.Read(b)
    if err != nil {
	return "", err
    }
    return base64.RawURLEncoding.EncodeToString(b), nil
}
```
This results in a token 43 characters in length, whereas previously, the JWT token was 116 characters long. Do we have a preferred character length and function for generating the token?

MacRae's package uses gorilla's `securecookie` to generate 64 chars:
```golang
func generateSessionKey() (string, error) {
    secureBytes := securecookie.GenerateRandomKey(32)
    if secureBytes == nil {
        return "", errors.New("Failed to generate random data for a key")
    }

    secureString := hex.EncodeToString(secureBytes)

    return secureString, nil

}
```

chrisgilmerproj commented:
@monfresh - sounds like 64 is a good limit. The JWT length I think was determined by the data we put in it. But I would agree that the more characters we use the more secure we are (probably).

monfresh commented:
I did some digging around and found that [OWASP recommends at least 64 bits of entropy](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Session_Management_Cheat_Sheet.md#session-id-length).

Python recommends [32 bytes of randomness](https://docs.python.org/3/library/secrets.html#how-many-bytes-should-tokens-use) (256 bits) to prevent brute force attacks, based on computing powers as of 2015. This means 43 characters when using [a-z][A-Z][0-9] and other URL-safe characters.

Rails has been using 32 characters via `SecureRandom.hex(16)` since at least 2009.
Rack defaults to 64 characters via `SecureRandom.hex(32)` since 2011.

According to [Wikipedia ](https://en.wikipedia.org/wiki/Brute-force_attack), "fifty supercomputers that could check a billion billion (10<sup>18</sup>) AES keys per second (if such a device could ever be made) would, in theory, require about 3×10<sup>51</sup> years to exhaust the 256-bit key space.

So, for now, I think we are safe with 32 bytes (256 bits), which is what we have right now with 43 characters by applying `base64.RawURLEncoding.EncodeToString`. In the future, we could update the `scs` package to allow the key generator to be configurable, as well as the key length.

monfresh commented:
@chrisgilmerproj I updated the Redis config here: https://github.com/transcom/mymove/pull/3926/commits/a95f98ff5c9f0ea26598a5227513f6f513b6d19a. I think I addressed all your comments. Next I'll work on the `/health` endpoint.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,153 @@
+	RedisDBNameFlag string = "redis-db-name"
+	// RedisConnectTimeoutFlag specifies how long to wait to establish a
+	// connection to the Redis instance
+	RedisConnectTimeoutFlag string = "redis-connect-timeout-in-seconds"
```
Can you add another param called `redis-enabled` that allows us to not connect to the redis backend? The mTLS stuff won't need Redis (at least for now) so I'd like to ensure all the `config/env/*.app-client-tls.env` files have `REDIS_ENABLED=0`. 

Similarly, can you add this block to all the `config/env/*.app.env` files?

```
REDIS_USER=...
REDIS_PORT=...
```

We'll use some fancy features of the ECS task definition to inject the REDIS_PASSWORD and REDIS_HOST from AWS SSM when we deploy, so no need to worry there.

monfresh replied:
Done here: https://github.com/transcom/mymove/pull/3926/commits/2404a0959f780f813db006187854648b1f9181aa

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,167 @@
+	redisConnectTimeout := v.GetInt(RedisConnectTimeoutFlag)
+	timeoutDuration := time.Duration(redisConnectTimeout) * time.Second
+
+	s := "redis://%s:%s@%s:%s/%s"
```
It turns out that username is optional. When using the Auth Token mechanism which is employed by AWS Elasticache we are only using a password. That would modify the auth here to look like:

```suggestion
	s := "redis://%s@%s:%s/%s"
```

So I'd add in a check here and if the username, password, or both are missing then the uri template should be changed.

Also, can you modify the variable name to be `redisUriTemplate` instead of `s`?

chrisgilmerproj commented:
I have put the new experimental redis host name in chamber under the name `redis_host`. You can see it with `chamber read app-experimental redis_host` if you'd like. The auth token is similarly added as `redis_password`. You will want to deploy this branch to experimental to see if you can connect to redis and then debug from there.

Additionally, until we complete the minor mod paperwork with USTC we won't be able to merge this PR so I'm going to leave it blocked until that paperwork is finished.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,174 @@
+
+	pool := &redis.Pool{
+		MaxIdle:     10,
+		IdleTimeout: 240 * time.Second,
```
We should be able to set the `MaxIdle` and `IdleTimeout` via CLI flags.

chrisgilmerproj commented on pkg/cli/redis.go:
```diff
@@ -0,0 +1,174 @@
+				"tcp",
+				redisURL,
+				redis.DialDatabase(redisDBName),
+				redis.DialPassword(redisPassword),
```
I noticed there's no connection timeout here, I'll add it.

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -479,10 +529,36 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+	gob.Register(auth.Session{})
+	var sessionManager *scs.SessionManager
+	sessionManager = scs.New()
+	sessionManager.Store = redisstore.New(redisPool)
```
I'm thinking that if Redis is not enabled then we shouldn't enable any of the session manager code. Those two things need to be tied together.

chrisgilmerproj commented on Makefile:
```diff
@@ -431,6 +435,34 @@ server_test_docker_down:
+
+.PHONY: redis_run
+redis_run: ## Run Redis
+	@echo "Starting the ${REDIS_DOCKER_CONTAINER} docker redis container..."
```
```
ifndef CIRCLECI
	brew services stop redis 2> /dev/null || true
endif
```

chrisgilmerproj commented on pkg/cli/session.go:
```diff
@@ -0,0 +1,68 @@
+package cli
```
You'll also need `pkg/cli/session_test.go`. Just copy the one from `redis_test.go`.

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -479,10 +529,36 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+	// IdleTimeout controls the maximum length of time a session can be inactive
+	// before it expires. The default is 15 minutes. To disable idle timeout in
+	// a non-production environment, set SESSION_IDLE_TIMEOUT_IN_MINUTES to 0.
+	idleTimeout := v.GetDuration(cli.SessionIdleTimeoutInMinutesFlag) * time.Minute
```
```suggestion
	idleTimeout := v.GetDuration(cli.SessionIdleTimeoutInMinutesFlag)
```

monfresh replied:
If I remove the `* time.Minute`, and if the ENV var is set to `2`, it interprets it as 2 nanoseconds. I tried setting the ENV var to `2m` to specify minutes, but that didn't work. Looking at the initialization, I would expect that any integer in the ENV var would be converted to minutes, but it doesn't look like it's doing that:
```
func InitSessionFlags(flag *pflag.FlagSet) {
	flag.Duration(SessionIdleTimeoutInMinutesFlag, (time.Duration(SessionIdleTimeoutInMinutes) * time.Minute), "Session idle timeout in minutes")
	flag.Duration(SessionLifetimeInHoursFlag, (time.Duration(SessionLifetimeInHours) * time.Hour), "Session absolute expiry in hours")
}
```

chrisgilmerproj replied:
It ought to work, I'm surprised. Looking at the code its pretty straight forward that if you give `2m` it ought to return a 2 minute duration.

https://github.com/spf13/pflag/blob/master/duration.go#L28

monfresh replied:
To close the loop on this, I just tried again, and setting the env var to `2m` does work. However, if you don't set the env var, the default value of 15 will be interpreted as 15ns, not 15 minutes. I also don't think we should rely on humans to type the correct letter after the number. I would prefer to enforce the correct time unit in the code and have folks enter only a number for the env var.

chrisgilmerproj replied:
My suggestion then is to just make the test check that no duration values are under 1s.

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -479,10 +529,36 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+	// before it expires. The lifetime is an 'absolute expiry' which is set when
+	// the session is first created or renewed (such as when a user signs in)
+	// and does not change. The default value is 24 hours.
+	lifetime := v.GetDuration(cli.SessionLifetimeInHoursFlag) * time.Hour
```
```suggestion
	lifetime := v.GetDuration(cli.SessionLifetimeInHoursFlag)
```

chrisgilmerproj commented:
@monfresh - I believe this is working now. A couple things to note.

For setup:

```
make redis_pull
make redis_run
```

For the redis-cli:

```
redis-dev
```

All of the targets in the `Makefile` that need `redis_reset` target are updated.

TLS is configured for use in AWS. We set a minimum TLS version of 1.2 so we stay in line with our security requirements.

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -479,10 +529,59 @@ func serveFunction(cmd *cobra.Command, args []string) error {
+		officeSession.Cookie.Secure = true
+	}
+
+	sessionManagers := [3]*scs.SessionManager{milSession, adminSession, officeSession}
```
Is it possible to setup all the session logic into a function that returns `sessionManagers`? Then we could put that code in `pkg/auth/session.go`, name it `SessionSetup` (or whatever) and do unit testing against it?

monfresh replied:
Absolutely. That was my plan as a follow up. I wanted to make sure things worked first, and then refactor. Putting all the app sessions in an array was the first thing that came to my mind. What do you think of that approach?

chrisgilmerproj replied:
I like it but it might cause some issues later if folks misuse the wrong element in the array. Maybe a struct with attributes like `GetAdminSession()` would be a good idea?

monfresh commented:
@chrisgilmerproj Should we add `redis_pull` as part of `make prereqs`?

chrisgilmerproj commented:
> @chrisgilmerproj Should we add `redis_pull` as part of `make prereqs`?

That seems like a great idea! I also wanted to make a `db_pull` target at some point but didn't want to add it to your PR. It never occurred to me that we could do `docker pull` as a target in `Makefile`s and its really helpful.

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-2321">MB-2321</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 6ff8ac7580608d63d16247b167f0c9128e0ddfec
</p>

chrisgilmerproj commented:
Sorry for the delay on approval. If you're working with someone on infrasec and the on-call team to merge this then go ahead. You've done great work on this and I'm excited to see MilMove using it.
